### ISSUE TITLE
java.lang.IllegalArgumentException Invalid ClientHandle

### ISSUE THREAD

**Issue by amglhit** [eclipse/paho.mqtt.android#318]
Invalid ClientHandle

org.eclipse.paho.android.service.MqttService.org.eclipse.paho.android.service.MqttConnection getConnection(java.lang.String)(SourceFile:588)
org.eclipse.paho.android.service.MqttService.void connect(java.lang.String,org.eclipse.paho.client.mqttv3.MqttConnectOptions,java.lang.String,java.lang.String)(SourceFile:328)
org.eclipse.paho.android.service.MqttAndroidClient.void doConnect()(SourceFile:467)
org.eclipse.paho.android.service.MqttAndroidClient.org.eclipse.paho.android.service.MqttService access$002(org.eclipse.paho.android.service.MqttAndroidClient,org.eclipse.paho.android.service.MqttService)(SourceFile:76)

**Comment by chintan-mishra** [comment]
Please reformat your snippet. And show us some code that you are using to connect with the broker.

**Comment by kuailewangzi0328** [comment]
you should not throw IllegalArgumentException in getConnection method…… we can not catch this exception in main thread, it make my application crash many times……

**Comment by chintan-mishra** [comment]
@kuailewangzi0328 you should not be doing any network operation on Main Thread of android. This will lead to crash and ANRs.

**Comment by kuailewangzi0328** [comment]
@chintan-mishra  yes,I have not do any network  operate in main thread, but I need call  mClient.connect(mConnectOptions,null, mConnectCallback)  in main thread, but in this method , you called pool.excute(doConnect();), and throw a IllegalArgumentException in code "getConnection(clientHandle)", so this exception throwed in another thread, then my app crashed because I cant catch this exception when I call "mClient.connect()" in main thread, how can I do?

**Comment by kuailewangzi0328** [comment]
[https://github.com/eclipse/paho.mqtt.android/blob/b5d5e5ef91e846d39a365b8c561fa9670d34f691/org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/MqttService.java#L588](url)

[https://github.com/eclipse/paho.mqtt.android/blob/b5d5e5ef91e846d39a365b8c561fa9670d34f691/org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/MqttAndroidClient.java#L431](url)

**Comment by chintan-mishra** [comment]
`IllegalArgumentException` is the correct exception but the issue is that you were not able to catch it on the main thread.

For now you can try this
```java
try {
  // Your code
} catch (MqttException | IllegalArgumentException e) {
  // More code to handle exception
}
```

**Comment by kuailewangzi0328** [comment]
I call mClient.connect() in A thread, but actually your code excuted and throw exception in B thread because your code excute by pool.excute() , so I cant catch any exception in my code ,do you understand?

**Comment by chintan-mishra** [comment]
Now I understand what problem you are facing.

**Comment by chintan-mishra** [comment]
What do I do to reproduce the error @kuailewangzi0328 ?

I tried connecting a vanilla application it seems to work fine/

**Comment by kuailewangzi0328** [comment]
In my application, I register a network state change broadcast  receiver ,when network disable, I disconnecte mqtt connection, and when network available, I will try to reconnect to the mqtt server, so when turn on/off my cellphone's wifi or celluar data, the code will be throw this exception. 
Actually in the begin, I find the exception throwed when I call mCliend.isconnected(), so I catch this exception. but then I find it crash as before, so when I read the source cod I find I cant catch the cxception when I call mClient.connect() because it run in another thread.

**Comment by kuailewangzi0328** [comment]
in my opinion, you should not just throw a exception when you get a null object in getConnection method,  in especial you excute it in other thread , it make my whole application crash again and again, even I cant catch it . you should invoke the callback onfail method, then I can make a try to reconnect or do other  thing …… I hope you can give me a solution to avoid my application crash…… thanks a lot

**Comment by wll995806658** [comment]
I also face this problem. Similar to his situation.I try to connect and disconnect frequently.It seems that such a problem occurs when I call disconnect before getting the callback result of the subscribe, but I'm not sure.

**Comment by ArcticSnowSky** [comment]
Heya @all,
as I am working on a project with this lib too, I got also contact with this problem.

I always got the problem when my app got restarted, however i have to add that i have a foreground_service keeping various connectionSystems (Bluetooth, Mqtt, DataRecorder).
The problem appeared when:
App gets closed (with isFinally() true), unregisterResources gets called. Inside, unregisterReceiver gets called. ClientHandle can't be set to null anymore.
When the app gets opened again,client Object I have is still the same (as it was stored in the foreground_service).
registerResources didn't help, because the ClientHandle inside was not null, on the next operation it will say InvalidClientHandle.

In my opinion, unregisterReceiver maybe shall set ClientHandle null, or registerReceiver might check ClientHandle validity and reset it if required?

**Comment by bailyzheng** [comment]
I also got this problem. It disappeared when I set the auto connection option to false.  Reconnect by some other method when it disconnect. You guys can try this method to get rid of this bug.
